### PR TITLE
feat: Calculate cost for models with "char without space" #154

### PR DESCRIPTION
**Pull request by astsiapanay**

### CODE REVIEW

File: src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java [hunk 1]
```diff
@@ -297,6 +298,11 @@ void handleResponse() {
                 context.setTokenUsage(tokenUsage);
                 proxy.getRateLimiter().increase(context);
                 tokenUsageFuture = Future.succeededFuture(tokenUsage);
+                try {
+                    tokenUsage.setCost(ModelCostCalculator.calculate(context));
+                } catch (Throwable e) {
+                    log.warn("Failed to calculate cost for model={}", context.getDeployment().getName());
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java]: Add trace and span.

File: src/main/java/com/epam/aidial/core/util/ModelCostCalculator.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            return cost;
+        }
+        return null;
+    }
+
+    private static int getResponseContentLength(ModelType modelType, Buffer responseBody) {
+        if (modelType == ModelType.EMBEDDING) {
+            return 0;
+        }
+        try (InputStream stream = new ByteBufInputStream(responseBody.getByteBuf())) {
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/util/ModelCostCalculator.java]: How does it work with streaming?

File: src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java (hunk 1 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java]: fixed

File: src/main/java/com/epam/aidial/core/util/ModelCostCalculator.java (hunk 2 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/util/ModelCostCalculator.java:79]: `DeploymentPostController` sets responseBody to a buffer when the stream is completed.
see com.epam.aidial.core.controller.DeploymentPostController#handleProxyResponse

Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/util/ModelCostCalculator.java]: fixed

**Review by artsiomkorzun — APPROVED** [review]

**Pull request merged**